File: cf-workers/src/push-http.ts
import type { ProviderConfig, PushResult } from "./types";
import { fetchWithTimeout } from "./rocom-client";
import { redactSensitiveText } from "./push-redaction";

export function jsonResult(
  payload: Record<string, unknown>,
  successCodes: Set<unknown>
): { success: boolean; message: string } {
  const code = payload.code ?? payload.errcode;
  let success = successCodes.has(code);
  if (code === undefined && Object.keys(payload).length === 0) {
    success = true;
  }
  const message = String(
    payload.message || payload.msg || payload.errmsg || JSON.stringify(payload)
  );
  return { success, message };
}

export async function readResponsePayload(resp: Response): Promise<{
  payload: Record<string, unknown>;
  text: string;
}> {
  const text = await resp.text();
  if (!text.trim()) return { payload: {}, text: "" };
  try {
    const parsed = JSON.parse(text) as unknown;
    if (parsed && typeof parsed === "object" && !Array.isArray(parsed)) {
      return { payload: parsed as Record<string, unknown>, text };
    }
  } catch {
    // non-json response body
  }
  return { payload: {}, text };
}

export function resultFromParsedResponse(
  provider: ProviderConfig,
  resp: Response,
  payload: Record<string, unknown>,
  text: string,
  successCodes: Set<unknown>
): PushResult {
  let { success, message } = jsonResult(payload, successCodes);
  const textMessage = text.slice(0, 200);
  if (Object.keys(payload).length === 0 && textMessage) {
    message = textMessage;
  }
  if (resp.status >= 400) {
    success = false;
    message = textMessage || message;
  }
  return {
    providerId: provider.id,
    providerName: provider.name,
    providerType: provider.type,
    success,
    message: redactSensitiveText(provider, message),
    statusCode: resp.status,
  };
}

export function providerErrorResult(provider: ProviderConfig, err: unknown): PushResult {
  return {
    providerId: provider.id,
    providerName: provider.name,
    providerType: provider.type,
    success: false,
    message: redactSensitiveText(provider, String(err)),
    statusCode: null,
  };
}

export async function postJson(
  provider: ProviderConfig,
  url: string,
  payload: Record<string, unknown>,
  timeoutSec: number,
  options?: { headers?: Record<string, string>; successCodes?: Set<unknown> }
): Promise<PushResult> {
  const successCodes = options?.successCodes ?? new Set([0, "0"]);
  try {
    const resp = await fetchWithTimeout(
      url,
      {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
          ...options?.headers,
        },
        body: JSON.stringify(payload),
      },
      timeoutSec
    );

    const { payload: respPayload, text } = await readResponsePayload(resp);
    return resultFromParsedResponse(
      provider,
      resp,
      respPayload,
      text,
      successCodes
    );
  } catch (err) {
    return providerErrorResult(provider, err);
  }
}

File: src/roco_push_console/push_http.py
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any

import requests

from .push_models import ProviderConfig, PushResult
from .push_redaction import redact_sensitive_text

HttpSession = requests.Session


@dataclass(frozen=True)
class JsonPostRequest:
    provider: ProviderConfig
    session: HttpSession
    url: str
    payload: dict[str, Any]
    timeout: int
    headers: dict[str, str] | None = None
    success_codes: set[Any] = field(default_factory=lambda: {0, "0"})


def json_result(payload: dict[str, Any], success_codes: set[Any]) -> tuple[bool, str]:
    code = payload.get("code", payload.get("errcode"))
    success = code in success_codes
    if code is None and not payload:
        success = True
    message = str(payload.get("message") or payload.get("msg") or payload.get("errmsg") or payload)
    return success, message


def result_from_response(
    provider: ProviderConfig,
    response: requests.Response,
    *,
    success_codes: set[Any],
) -> PushResult:
    try:
        payload = response.json()
    except ValueError:
        payload = {}
    success, message = json_result(payload, success_codes)
    if response.status_code >= 400:
        success = False
        message = response.text[:200] or message
    return PushResult(
        provider.id,
        provider.name,
        provider.type,
        success,
        redact_sensitive_text(provider, message),
        response.status_code,
    )


def post_json(request: JsonPostRequest) -> PushResult:
    response = request.session.post(
        request.url,
        json=request.payload,
        headers=request.headers,
        timeout=request.timeout,
    )
    return result_from_response(
        request.provider,
        response,
        success_codes=request.success_codes,
    )
